ev3dev2/motor.py

```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
+                    kp=11.3, ki=0.05, kd=3.2,
+                    speed=SpeedPercent(30),
+                    follow_for=follow_for_ms,
+                    ms=4500
+                )
+            except Exception:
+                tank.stop()
+                raise
+        """
+        assert self.gyro, "GyroSensor must be defined"
+
+        integral = 0.0
+        last_error = 0.0
+        derivative = 0.0
+        speed_native_units = speed.to_native_units(self.left_motor)
+        MAX_SPEED = SpeedNativeUnits(self.max_speed)
+
+        while follow_for(self, **kwargs):
+            current_angle = self.gyro.angle
+            error = current_angle - target_angle
+            integral = integral + error
+            derivative = error - last_error
+            last_error = error
+            turn_native_units = (kp * error) + (ki * integral) + (kd * derivative)
+
+            left_speed = SpeedNativeUnits(speed_native_units - turn_native_units)
+            right_speed = SpeedNativeUnits(speed_native_units + turn_native_units)
+
+            if left_speed > MAX_SPEED:
+                log.info("%s: left_speed %s is greater than MAX_SPEED %s" %
+                        (self, left_speed, MAX_SPEED))
+                self.stop()
+                raise FollowGyroAngleErrorTooFast(
+                    "The robot is moving too fast to follow the angle")
+
+            if right_speed > MAX_SPEED:
+                log.info("%s: right_speed %s is greater than MAX_SPEED %s" %
+                        (self, right_speed, MAX_SPEED))
+                self.stop()
+                raise FollowGyroAngleErrorTooFast(
+                    "The robot is moving too fast to follow the angle")
+
+            if sleep_time:
+                time.sleep(sleep_time)
+
+            self.on(left_speed, right_speed)
+
+        self.stop()
+
+    def pivot_gyro(self,
+            speed,
+            target_angle=0,
+            sleep_time=0.01
+        ):
+        """
+        Pivot Turn
+
+        ``speed`` is the desired speed of the midpoint of the robot
+
+        ``target_angle`` is the angle we want drive
+
+        ``sleep_time`` is how many seconds we sleep on each pass through
+            the loop.  This is to give the robot a chance to react
+            to the new motor settings. This should be something small such
+            as 0.01 (10ms).
+
+        Example:
+
+        .. code:: python
+        
+            from ev3dev2.motor import OUTPUT_A, OUTPUT_B, MoveTank, SpeedPercent
+            from ev3dev2.sensor.lego import GyroSensor
+            tank = MoveTank(OUTPUT_A, OUTPUT_B)
+            tank.gyro = GyroSensor()
+            try:
+                # Reset gyro sensor to zero
+                tank.calibrate_gyro()
+
+                # Pivot 30 degrees
+                tank.pivot_gyro(
+                    speed=SpeedPercent(5),
+                    target_angle(30)
+                )
+            except Exception:
```

Comment:
same note as above regarding exception type

Reply:
I'll remove. This method doesn't throw an exception unless someone forgot to initialize the gyro. But this sample properly initializes it so there will be zero chance of the AssertionError being raised.